lib/connection_pool.rb

```diff
@@ -27,34 +29,66 @@
 #      $redis.lpop('my-list') if $redis.llen('my-list') > 0
 #    end
 #
+# Note that there's no way to pass a disconnection function to this
+# usage, nor any way to guarantee that subsequent calls will go to the
+# same connection (if your connection has any concept of sessions, this
+# may be important). We strongly recommend against using wrapped
+# connections in production environments.
+#
 # Accepts the following options:
 # - :size - number of connections to pool, defaults to 5
 # - :timeout - amount of time to wait for a connection if none currently available, defaults to 5 seconds
+# - :max_age - maximum number of seconds that a connection may be alive for (will recycle on checkin/checkout)
+# - :connect - callable for creating a connection
+# - :disconnect - callable for shutting down a connection
 #
 class ConnectionPool
-  DEFAULTS = {size: 5, timeout: 5}
-
-  class Error < RuntimeError
-  end
+  DEFAULTS = {size: 5, timeout: 1, max_age: Float::INFINITY}
 
   def self.wrap(options, &block)
-    Wrapper.new(options, &block)
+    if block_given?
+      options[:connect] = block
+    end
+    Wrapper.new(options)
   end
 
   def initialize(options = {}, &block)
-    raise ArgumentError, 'Connection pool requires a block' unless block
-
     options = DEFAULTS.merge(options)
 
     @size = options.fetch(:size)
     @timeout = options.fetch(:timeout)
+    @max_age = options.fetch(:max_age).to_f
+
+    if @max_age <= 0
+      raise ArgumentError.new(":max_age must be > 0")
+    end
+
+    if block_given?
+      if options.include?(:connect)
+        raise ArgumentError.new("Block passed to ConnectionPool *and* :connect in options")
+      else
+        options[:connect] = block
+      end
+    end
 
-    @available = TimedStack.new(@size, &block)
+    @manager = ConnectionPool::ConnectionManager.new(options[:connect], options[:disconnect])
```

Comment:
I would suggest to keep methods names consistent. Use `connect_with` instead of `connect` and `connect_proc`
Same for `disconnect`